### ISSUE TITLE
Symlink to now.json

### ISSUE THREAD

**Issue by samuelcastro** [UnlyEd/next-right-now#182]
To avoid rely on manually create a symlink of `now.json` from `now.*.staging.json` and `now.*.production.json` based on the environment, it'd be way easier to symlink it directly from our CI/CD workflow. 

I was having a problem on production by the simple fact I forgot to create the symlink manually.

Thoughts?

**Comment by Vadorequest** [comment]
So, instead of having a now.json file (or vercel.json), we'd have no file, but the CI/CD process would create the file itself?

I can see a few issues with this.
- The now.json file is currently used to resolve which customer should be deployed by CI/CD, it's basically the entry point of all CI/CD actions. (`CUSTOMER_TO_DEPLOY="${MANUAL_TRIGGER_CUSTOMER:-$(cat now.json | jq -r '.build.env.NEXT_PUBLIC_CUSTOMER_REF')}"`)
- The now.json file is still needed when using the `now` cli. For deployments (e.g: `yarn:deploy:customer1`), we provide the now.*.json file to use, but other actions might fail if no now.json file is present

So:
- How would the CI/CD know what customer should be deployed? (which now.*.json file to use)


On the bright side, I've always wondered how we could improve this workflow, and maybe deploy several customers at once. I'm not sure we'll find a solution, but that's worth thinking about it.

**Comment by samuelcastro** [comment]
We'd know that from `NEXT_PUBLIC_CUSTOMER_REF`, we don't need the now.json to know that.

**Comment by Vadorequest** [comment]
While this is true in a local environment, it's not how it works when deploying through GitHub Actions. 

Do you mean we could read the .env file that contains the `NEXT_PUBLIC_CUSTOMER_REF` when deploying through GitHub Actions? And then, resolve which now.*.json file to use based on the value of `NEXT_PUBLIC_CUSTOMER_REF`? That could work. 🤔

**Comment by samuelcastro** [comment]
Exactly. Via https://github.com/marketplace/actions/environment-variables-from-dotenv for instance.

**Comment by Vadorequest** [comment]
On second thought, I might have misunderstood your pain point here.

> I was having a problem on production by the simple fact I forgot to create the symlink manually.

What do you mean? Do you change the symlink regularly? Because I don't, the symlink is meant to make a choice between `customers`, not to deploy to production. Deploying to production is automated when pushing into the master branch. Maybe your workflow is different from the one I intended to be used?

**Comment by samuelcastro** [comment]
I don't change the symlink regularly, the fact is that this is a manual process that could be easily automated. The problem I had was that on dev stage the symlink was pointing to *.staging.json and when I merged to master I forgot to update it accordingly.

**Comment by samuelcastro** [comment]
But setup is exactly the same you're using, although I'm using ST rather than MST.

**Comment by Vadorequest** [comment]
There is still something odd here, because on my setup the symlink always point to one of the now.*.staging.json even when I merge it into the master branch, but it deploys using the production file.

Maybe something broke when you migrated from MST to ST? 

- Are you sure your Github Actions triggers the "deploy-zeit-production.yml" workflow when you merge a commit onto master?
- Are you sure your "deploy-zeit-production.yml" "start-production-deployment" step deploys the production command? `yarn deploy:$CUSTOMER_TO_DEPLOY:production:simple --token $ZEIT_TOKEN`
- Are you sure the `yarn deploy:$CUSTOMER_TO_DEPLOY:production:simple` command uses a "production" now.json file? (E.g: `yarn now:cleanup && now --confirm --local-config=now.customer1.production.json --prod --debug`)

**Comment by samuelcastro** [comment]
Yes I'm sure, everything is fine, and it was actually even deploying however I noticed that i18n was not working, it was not pulling translations from locize, that's when I realized my `now.json` was not linked to `now.*.production.json`.

**Comment by Vadorequest** [comment]
Hum, my point is that now.*.production.json file should never be linked to now.json. Only the staging.

Could you share your package.json and workflow staging+prod?

**Comment by samuelcastro** [comment]
scripts on package.json
```json
"scripts": {
    "start": "next dev --port 8888",
    "start:tunnel": "ngrok http 8888",
    "build": "yarn test:once && next build",
    "build:profiler": "next build --profile",
    "analyse:bundle": "ANALYZE_BUNDLE=true yarn start",
    "analyse:unused": "next-unused",
    "svg": "npx svgr -d src/svg src/svg --ext tsx --template src/common/utils/svg/svgTemplate.ts",
    "deploy": "yarn deploy:app",
    "deploy:all": "yarn deploy:app",
    "deploy:all:production": "yarn deploy:app:production",
    "deploy:all:all": "yarn deploy:all && yarn deploy:all:production",
    "deploy:app:all": "yarn deploy:app && yarn deploy:app:production",
    "deploy:app": "yarn now:cleanup && now --confirm --local-config=now.app.staging.json --debug",
    "deploy:app:production": "yarn deploy:app:production:simple && yarn e2e:app:production",
    "deploy:app:production:simple": "yarn now:cleanup && now --confirm --local-config=now.app.production.json --prod --debug",
    "deploy:fake": "git commit --allow-empty -m \"Fake empty commit (force CI trigger)\"",
    "now:cleanup": "npx del-cli .now/ && npx del-cli .vercel/",
    "e2e": "yarn e2e:run",
    "e2e:all:production": "yarn e2e:app:production",
    "e2e:app:production": "CYPRESS_STAGE=production yarn e2e:run",
    "e2e:app:production:open": "CYPRESS_STAGE=production yarn e2e:open",
    "e2e:install": "cypress install",
    "e2e:open": "CYPRESS_STAGE=${CYPRESS_STAGE:-development}; cypress open --config-file cypress/config-$CYPRESS_STAGE.json",
    "e2e:run": "CYPRESS_STAGE=${CYPRESS_STAGE:-development}; cypress run --config-file cypress/config-$CYPRESS_STAGE.json",
    "e2e:run:spec:common": "CYPRESS_STAGE=${CYPRESS_STAGE:-development}; cypress run --config-file cypress/config-$CYPRESS_STAGE.json --spec 'cypress/integration/app/common/*.js'",
    "e2e:ci": "yarn e2e:install && cypress run --record",
    "preversion": "yarn lint:once && yarn test:once",
    "release": "yarn bump --commit --tag && git add CHANGELOG.md README.md && git commit --amend --no-edit && git push && git push --tags",
    "doc:start": "cd docs/ && bundle exec jekyll serve --config _config-development.yml",
    "doc:start:fast": "cd docs/ && bundle exec jekyll serve --config _config-development.yml --incremental",
    "doc:gem:install": "cd docs/ && bundle install",
    "lint": "esw src/ -w --ext .ts --ext .tsx",
    "lint:once": "eslint src/ --ext .ts --ext .tsx",
    "lint:fix": "eslint src/ --ext .ts --ext .tsx --fix",
    "lint:fix:preview": "eslint src/ --ext .ts --ext .tsx --fix-dry-run",
    "test": "NODE_ENV=test jest --watchAll",
    "test:group:api": "NODE_ENV=test jest --group=api --watchAll",
    "test:group:components": "NODE_ENV=test jest --group=components --watchAll",
    "test:group:integration": "NODE_ENV=test jest --group=integration --watchAll",
    "test:group:unit": "NODE_ENV=test jest --group=unit --watchAll",
    "test:group:utils": "NODE_ENV=test jest --group=utils --watchAll",
    "test:once": "NODE_ENV=test jest --runInBand --detectOpenHandles",
    "test:once:group:no-integration": "NODE_ENV=test jest --group=-integration",
    "test:coverage": "NODE_ENV=test jest --coverage",
    "test:coverage:group:no-integration": "NODE_ENV=test jest --group=-integration --coverage",
    "test:config": "NODE_ENV=test jest --showConfig",
    "security:audit": "yarn audit",
    "packages:upgrade": "yarn upgrade-interactive --latest",
    "codegen": "npx graphql-codegen",
    "add:local": "npx yalc link",
    "update:local": "npx yalc update"
  }
```
Workflow staging:
```yml
# Summary:
# Creates a new deployment on Zeit's platform, when anything is pushed in any branch (except for the "master" branch).
# Read ./README.md for extensive documentation

name: Deploy to Zeit (staging)

on:
  # There are several ways to trigger Github actions - See https://help.github.com/en/actions/reference/events-that-trigger-workflows#example-using-a-single-event for a comprehensive list:
  # - "push": Triggers each time a commit is pushed
  # - "pull_request": Triggers each time a commit is pushed within a pull request, it makes it much easier to write comments within the PR, but it suffers some strong limitations:
  #   - There is no way to trigger when a PR is merged into another - See https://github.community/t/pull-request-action-does-not-run-on-merge/16092?u=vadorequest
  #   - It won't trigger when the PR is conflicting with its base branch - See https://github.community/t/run-actions-on-pull-requests-with-merge-conflicts/17104/2?u=vadorequest
  push: # Triggers on each pushed commit
    branches-ignore:
      - 'master'

jobs:
  # Configures the deployment environment, install dependencies (like node, npm, etc.) that are requirements for the upcoming jobs
  # Ex: Necessary to run `yarn deploy`
  setup-environment:
    name: Setup deployment environment (Ubuntu 18.04 - Node 12.x)
    runs-on: ubuntu-18.04
    steps:
      - name: Installing node.js
        uses: actions/setup-node@v1 # Used to install node environment - XXX https://github.com/actions/setup-node
        with:
          node-version: '12.x' # Use the same node.js version as the one Zeit's uses (currently node12.x)

  # Starts a Zeit deployment, using the staging configuration file of the default institution
  # The default institution is the one defined in the `now.json` file (which is a symlink to the actual file)
  # N.B: It's Zeit that will perform the actual deployment
  start-staging-deployment:
    name: Starts Zeit deployment (staging) (Ubuntu 18.04)
    runs-on: ubuntu-18.04
    needs: setup-environment
    steps:
      - uses: actions/checkout@v1 # Get last commit pushed - XXX See https://github.com/actions/checkout
      - name: Deploying on Zeit (staging)
        # Workflow:
        #   - Get stdout from deploy command (stderr shows build steps and stdout shows final url, which is what we are really looking for)
        #   - Set the deployment url that will be included in the eventual PR comment
        #   - Create a branch alias and link it to the deployment (so that each branch has its own domain automatically aliased to the lastest commit)
        run: |
          # Print the version of the "now" CLI being used (helps debugging)
          now --version

          # Deploy the customer on Vercel using the customer specified in now.json (e.g: "yarn deploy:app")
          # Store the output in a variable so we can extract metadata from it
          ZEIT_DEPLOYMENT_OUTPUT=`yarn deploy:$(cat now.json | jq -r '.build.env.NEXT_PUBLIC_PROJECT_REF') --token $ZEIT_TOKEN`

          # Extract the Vercel deployment url from the deployment output
          ZEIT_DEPLOYMENT_URL=`echo $ZEIT_DEPLOYMENT_OUTPUT | egrep -o 'https?://[^ ]+.vercel.app'`
          echo "::set-env name=ZEIT_DEPLOYMENT_URL::$ZEIT_DEPLOYMENT_URL"
          echo "Deployment url: " $ZEIT_DEPLOYMENT_URL

          # Build the alias domain based on our own rules (NRN own rules)
          # Basically, if a branch starts with "v[0-9]" then we consider it to be a "preset branch" and we use the branch name as alias (e.g: "v2-mst-xxx")
          # Otherwise, we use the deployment name and suffix it using the branch name (e.g: "v2-mst-xxx-branch-name")
          if [[ ${CURRENT_BRANCH##*/} =~ ^v[0-9]{1,}- ]]; then # Checking if pattern matches with "vX-" where X is a number
            ZEIT_DEPLOYMENT_ALIAS=${CURRENT_BRANCH##*/}
          else
            ZEIT_DEPLOYMENT_ALIAS=$(cat now.json | jq -r '.name')-${CURRENT_BRANCH##*/}
          fi

          # Zeit alias only allows 41 characters in the domain name, so we only keep the first 41 (41 = 53 - 11 - 1) characters (because Zeit needs 7 chars for ".vercel.app" at the end of the domain name, and count starts at 1, not 0)
          # Also, in order to remove forbidden characters, we transform every characters which are not a "alnum" and \n or \r into '-'
          ZEIT_DEPLOYMENT_ALIAS=$(echo $ZEIT_DEPLOYMENT_ALIAS | head -c 41 | tr -c '[:alnum:]\r\n' - | tr '[:upper:]' '[:lower:]')

          # Recursively remove any trailing dash ('-') to protect against invalid alias domain (not allowed to end with a trailing slash)
          while [[ "$ZEIT_DEPLOYMENT_ALIAS" == *- ]]
          do
            ZEIT_DEPLOYMENT_ALIAS=${ZEIT_DEPLOYMENT_ALIAS::-1}
          done

          # Build alias url and create Vercel alias
          ZEIT_DEPLOYMENT_ALIAS=$ZEIT_DEPLOYMENT_ALIAS.vercel.app
          echo "::set-env name=ZEIT_DEPLOYMENT_ALIAS::https://$ZEIT_DEPLOYMENT_ALIAS"
          echo "Alias domain: "$ZEIT_DEPLOYMENT_ALIAS
          echo "Aliasing the deployment using the git branch alias:"
          echo "npx now alias "$ZEIT_DEPLOYMENT_URL $ZEIT_DEPLOYMENT_ALIAS
          npx now alias $ZEIT_DEPLOYMENT_URL $ZEIT_DEPLOYMENT_ALIAS --token $ZEIT_TOKEN
        env:
          ZEIT_TOKEN: ${{ secrets.ZEIT_TOKEN }} # Passing github's secret to the worker
          CURRENT_BRANCH: ${{ github.ref }} # Passing current branch to the worker

      # We need to find the PR id. Will be used later to comment on that PR.
      - name: Finding Pull Request ID
        uses: jwalton/gh-find-current-pr@v1
        id: pr_id_finder
        if: always() # It forces the job to be always executed, even if a previous job fail.
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}

      # On deployment failure, add a comment to the PR
      - name: Comment PR (Deployment failure)
        uses: peter-evans/create-or-update-comment@v1
        if: failure()
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          issue-number: ${{ steps.pr_id_finder.outputs.number }}
          body: |
            :x:&nbsp; Deployment **FAILED**
            Commit ${{ github.sha }} failed to deploy to ${{ env.ZEIT_DEPLOYMENT_URL }}
            [click to see logs](https://github.com/samuelcastro/app-dom/pull/${{ steps.pr_id_finder.outputs.number }}/checks)

      # On deployment success, add a comment to the PR
      - name: Comment PR (Deployment success)
        uses: peter-evans/create-or-update-comment@v1
        if: success()
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          issue-number: ${{ steps.pr_id_finder.outputs.number }}
          body: |
            :white_check_mark:&nbsp; Deployment **SUCCESS**
            Commit ${{ github.sha }} successfully deployed to ${{ env.ZEIT_DEPLOYMENT_URL }}
            Deployment aliased as ${{ env.ZEIT_DEPLOYMENT_ALIAS }}

  # Runs E2E tests against the Zeit deployment
  run-2e2-tests:
    name: Run end to end (E2E) tests (Ubuntu 18.04)
    runs-on: ubuntu-18.04
    # Docker image with Cypress pre-installed
    # https://github.com/cypress-io/cypress-docker-images/tree/master/included
    container: cypress/included:3.8.3
    needs: start-staging-deployment
    steps:
      - uses: actions/checkout@v1 # Get last commit pushed - XXX See https://github.com/actions/checkout
      - name: Resolving deployment url from Zeit
        # The following workflow is:
        #  - getting all deployments data (by using the scope in `now.json`)
        #  - then we get the last url (in Node.js it corresponds as `response.deployments[0].url`
        #  - and then we remove the `"` character to pre-format url
        # We need to set env the url for next step, formatted as `https://${url provided by API}`
        run: |
          apt update -y >/dev/null && apt install -y jq >/dev/null
          ZEIT_DEPLOYMENT=`curl -H 'Accept: application/json' -H 'Content-Type: application/json' -H 'Authorization: Bearer ${{ secrets.ZEIT_TOKEN }}' https://api.zeit.co/v5/now/deployments?teamId=$(cat now.json | jq -r '.scope') | jq '.deployments [0].url' | tr -d \"`
          echo "::set-env name=ZEIT_DEPLOYMENT_URL::https://$ZEIT_DEPLOYMENT"

      # Run the E2E tests against the new Zeit deployment
      - name: Run E2E tests (Cypress)
        uses: cypress-io/github-action@v1 # XXX See https://github.com/cypress-io/github-action
        with:
          wait-on: ${{ env.ZEIT_DEPLOYMENT_URL }} # Be sure that the endpoint is ready by pinging it before starting tests, it has a timeout of 60seconds
          config-file: cypress/config-development.json # The config file itself doesn't matter because we will override most settings anyway. We just need `projectId` to run the tests.
          config: baseUrl=${{ env.ZEIT_DEPLOYMENT_URL }} # Overriding baseUrl provided by config file to test the new deployment

      # On E2E failure, upload screenshots
      - name: Uplad screenshots artifacts (E2E failure)
        uses: actions/upload-artifact@v1 # On failure we upload artifacts, https://help.github.com/en/actions/automating-your-workflow-with-github-actions/persisting-workflow-data-using-artifacts
        if: failure()
        with:
          name: screenshots
          path: cypress/screenshots/

      # On E2E failure, upload videos
      - name: Uplad videos artifacts (E2E failure)
        uses: actions/upload-artifact@v1 # On failure we upload artifacts, https://help.github.com/en/actions/automating-your-workflow-with-github-actions/persisting-workflow-data-using-artifacts
        if: failure()
        with:
          name: videos
          path: cypress/videos/

      # We need to find the PR id. Will be used later to comment on that PR.
      - name: Finding Pull Request ID
        uses: jwalton/gh-find-current-pr@v1
        id: pr_id_finder
        if: always() # It forces the job to be always executed, even if a previous job fail.
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}

      # On E2E failure, add a comment to the PR with additional information
      - name: Comment PR (E2E failure)
        uses: peter-evans/create-or-update-comment@v1
        if: failure()
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          issue-number: ${{ steps.pr_id_finder.outputs.number }}
          body: |
            :x:&nbsp; E2E tests **FAILED**
            Download artifacts (screenshots + videos) from [`checks`](https://github.com/samuelcastro/app-dom/pull/${{ steps.pr_id_finder.outputs.number }}/checks) section

      # On E2E success, add a comment to the PR
      - name: Comment PR (E2E success)
        uses: peter-evans/create-or-update-comment@v1
        if: success()
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          issue-number: ${{ steps.pr_id_finder.outputs.number }}
          body: |
            :white_check_mark:&nbsp; E2E tests **SUCCESS**
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

  # Run Lighthouse reports in parallel of E2E tests
  run-lighthouse-tests:
    name: Run LightHouse checks (Ubuntu 18.04)
    runs-on: ubuntu-18.04
    needs: start-staging-deployment
    steps:
      - uses: actions/checkout@v1 # Get last commit pushed - XXX See https://github.com/actions/checkout
      - name: Resolving deployment url from Zeit
        # The following workflow is:
        #  - getting all deployments data (by using the scope in `now.json`)
        #  - then we get the last url (in Node.js it corresponds as `response.deployments[0].url`
        #  - and then we remove the `"` character to pre-format url
        # We need to set env the url for next step, formatted as `https://${url provided by API}/en` using /en endpoint to improve perfs by avoiding the url redirect on /
        run: |
          apt update -y >/dev/null && apt install -y jq >/dev/null
          ZEIT_DEPLOYMENT=`curl -H 'Accept: application/json' -H 'Content-Type: application/json' -H 'Authorization: Bearer ${{ secrets.ZEIT_TOKEN }}' https://api.zeit.co/v5/now/deployments?teamId=$(cat now.json | jq -r '.scope') | jq '.deployments [0].url' | tr -d \"`
          echo "::set-env name=ZEIT_DEPLOYMENT_URL::https://$ZEIT_DEPLOYMENT"
        env:
          ZEIT_TOKEN: ${{ secrets.ZEIT_TOKEN }} # Passing github's secret to the worker
      # In order to store reports and then upload it, we need to create the folder before any tests
      - name: Create temporary folder for artifacts storage
        run: mkdir /tmp/lighthouse-artifacts

      # It runs the lighthouse report for a provided url and create a HTML report in the specified directory
      # Action documentation: https://github.com/marketplace/actions/lighthouse-check#usage-standard-example
      - name: Run Lighthouse
        uses: foo-software/lighthouse-check-action@v2.0.5
        id: lighthouseCheck
        with: # See https://github.com/marketplace/actions/lighthouse-check#inputs for all options
          accessToken: ${{ secrets.GITHUB_TOKEN }} # Providing a token to comment the PR if you are using a "pull_request" as trigger, or using "commentUrl" option.
          commentUrl: https://api.github.com/repos/samuelcastro/app-dom-v2/commits/${{ github.sha }}/comments
          prCommentEnabled: true # Whether to comment on the PR (default: true). Disabled because we use our own "Comment PR (LightHouse report)" action
          prCommentSaveOld: true # If true, then add a new comment for each commit. Otherwise, update the latest comment (default: false).
          outputDirectory: /tmp/lighthouse-artifacts # Used to upload artifacts.
          emulatedFormFactor: all # Run LightHouse against "mobile", "desktop", or "all" devices
          urls: ${{ env.ZEIT_DEPLOYMENT_URL }}, ${{ env.ZEIT_DEPLOYMENT_URL }}/en, ${{ env.ZEIT_DEPLOYMENT_URL }}/fr, ${{ env.ZEIT_DEPLOYMENT_URL }}/en-US, ${{ env.ZEIT_DEPLOYMENT_URL }}/fr-FR
          locale: en

      # Upload HTML report create by lighthouse, could be useful
      - name: Upload artifacts
        uses: actions/upload-artifact@v1
        with:
          name: Lighthouse reports
          path: /tmp/lighthouse-artifacts

      # Using a pre-build action to make the action fail if your score is too low. It can be really interesting to track a low score on a commit
      # You can remove this action IF you don't want lighthouse to be a blocking point in your CI
      # Official documentation: https://github.com/foo-software/lighthouse-check-status-action
      - name: Handle Lighthouse Check results
        uses: foo-software/lighthouse-check-status-action@v1.0.1
        with:
          lighthouseCheckResults: ${{ steps.lighthouseCheck.outputs.lighthouseCheckResults }}
          minAccessibilityScore: "50"
          minBestPracticesScore: "0"
          minPerformanceScore: "10"
          minProgressiveWebAppScore: "50"
          minSeoScore: "50"
```

Workflow production:
```yml
# Summary:
# Creates a new deployment on Zeit's platform, when anything is pushed in any branch (except for the "master" branch).
# Read ./README.md for extensive documentation

name: Deploy to Zeit (production)

on:
  # There are several ways to trigger Github actions - See https://help.github.com/en/actions/reference/events-that-trigger-workflows#example-using-a-single-event for a comprehensive list:
  # - "push": Triggers each time a commit is pushed
  # - "pull_request": Triggers each time a commit is pushed within a pull request, it makes it much easier to write comments within the PR, but it suffers some strong limitations:
  #   - There is no way to trigger when a PR is merged into another - See https://github.community/t/pull-request-action-does-not-run-on-merge/16092?u=vadorequest
  #   - It won't trigger when the PR is conflicting with its base branch - See https://github.community/t/run-actions-on-pull-requests-with-merge-conflicts/17104/2?u=vadorequest
  push: # Triggers on each pushed commit
    branches:
      - 'master'

jobs:
  # Configures the deployment environment, install dependencies (like node, npm, etc.) that are requirements for the upcoming jobs
  # Ex: Necessary to run `yarn deploy`
  setup-environment:
    name: Setup deployment environment (Ubuntu 18.04 - Node 12.x)
    runs-on: ubuntu-18.04
    steps:
      - name: Installing node.js
        uses: actions/setup-node@v1 # Used to install node environment - XXX https://github.com/actions/setup-node
        with:
          node-version: '12.x' # Use the same node.js version as the one Zeit's uses (currently node12.x)

  # Starts a Zeit deployment, using the production configuration file of the default institution
  # The default institution is the one defined in the `now.json` file (which is a symlink to the actual file)
  # N.B: It's Zeit that will perform the actual deployment
  start-production-deployment:
    name: Starts Zeit deployment (production) (Ubuntu 18.04)
    runs-on: ubuntu-18.04
    needs: setup-environment
    steps:
      - uses: actions/checkout@v1 # Get last commit pushed - XXX See https://github.com/actions/checkout
      - name: Deploying on Zeit (production)
        # Workflow:
        # Deploy the project in production
        run: |
          # Print the version of the "now" CLI being used (helps debugging)
          now --version

          # Deploy the project on Vercel using the project specified in now.json (e.g: "yarn deploy:app:production:simple")
          yarn deploy:$(cat now.json | jq -r '.build.env.NEXT_PUBLIC_PROJECT_REF'):production:simple --token $ZEIT_TOKEN
        env:
          ZEIT_TOKEN: ${{ secrets.ZEIT_TOKEN }} # Passing github's secret to the worker

  # Runs E2E tests against the Zeit deployment
  run-2e2-tests:
    name: Run end to end (E2E) tests (Ubuntu 18.04)
    runs-on: ubuntu-18.04
    # Docker image with Cypress pre-installed
    # https://github.com/cypress-io/cypress-docker-images/tree/master/included
    container: cypress/included:3.8.3
    needs: start-production-deployment
    steps:
      - uses: actions/checkout@v1 # Get last commit pushed - XXX See https://github.com/actions/checkout
      - name: Resolving deployment url from Zeit
        # The following workflow is:
        #  - getting all deployments data (by using the scope in `now.json`)
        #  - then we get the last url (in Node.js it corresponds as `response.deployments[0].url`
        #  - and then we remove the `"` character to pre-format url
        # We need to set env the url for next step, formatted as `https://${url provided by API}`
        run: |
          apt update -y >/dev/null && apt install -y jq >/dev/null
          ZEIT_DEPLOYMENT=`curl -H 'Accept: application/json' -H 'Content-Type: application/json' -H 'Authorization: Bearer ${{ secrets.ZEIT_TOKEN }}' https://api.zeit.co/v5/now/deployments?teamId=$(cat now.json | jq -r '.scope') | jq '.deployments [0].url' | tr -d \"`
          echo "::set-env name=ZEIT_DEPLOYMENT_URL::https://$ZEIT_DEPLOYMENT"

      # Run the E2E tests against the new Zeit deployment
      - name: Run E2E tests (Cypress)
        uses: cypress-io/github-action@v1 # XXX See https://github.com/cypress-io/github-action
        with:
          wait-on: ${{ env.ZEIT_DEPLOYMENT_URL }} # Be sure that the endpoint is ready by pinging it before starting tests, it has a timeout of 60seconds
          config-file: cypress/config-production.json # The config file itself doesn't matter because we will override most settings anyway. We just need `projectId` to run the tests.
          config: baseUrl=${{ env.ZEIT_DEPLOYMENT_URL }} # Overriding baseUrl provided by config file to test the new deployment

      # On E2E failure, upload screenshots
      - name: Uplad screenshots artifacts (E2E failure)
        uses: actions/upload-artifact@v1 # On failure we upload artifacts, https://help.github.com/en/actions/automating-your-workflow-with-github-actions/persisting-workflow-data-using-artifacts
        if: failure()
        with:
          name: screenshots
          path: cypress/screenshots/

      # On E2E failure, upload videos
      - name: Uplad videos artifacts (E2E failure)
        uses: actions/upload-artifact@v1 # On failure we upload artifacts, https://help.github.com/en/actions/automating-your-workflow-with-github-actions/persisting-workflow-data-using-artifacts
        if: failure()
        with:
          name: videos
          path: cypress/videos/

  # Run Lighthouse reports in parallel of E2E tests
  run-lighthouse-tests:
    name: Run LightHouse checks (Ubuntu 18.04)
    runs-on: ubuntu-18.04
    needs: start-production-deployment
    steps:
      - uses: actions/checkout@v1 # Get last commit pushed - XXX See https://github.com/actions/checkout
      - name: Resolving deployment url from Zeit
        # The following workflow is:
        #  - getting all deployments data (by using the scope in `now.json`)
        #  - then we get the last url (in Node.js it corresponds as `response.deployments[0].url`
        #  - and then we remove the `"` character to pre-format url
        # We need to set env the url for next step, formatted as `https://${url provided by API}/en` using /en endpoint to improve perfs by avoiding the url redirect on /
        run: |
          apt update -y >/dev/null && apt install -y jq >/dev/null
          ZEIT_DEPLOYMENT=`curl -H 'Accept: application/json' -H 'Content-Type: application/json' -H 'Authorization: Bearer ${{ secrets.ZEIT_TOKEN }}' https://api.zeit.co/v5/now/deployments?teamId=$(cat now.json | jq -r '.scope') | jq '.deployments [0].url' | tr -d \"`
          echo "::set-env name=ZEIT_DEPLOYMENT_URL::https://$ZEIT_DEPLOYMENT"
        env:
          ZEIT_TOKEN: ${{ secrets.ZEIT_TOKEN }} # Passing github's secret to the worker
      # In order to store reports and then upload it, we need to create the folder before any tests
      - name: Create temporary folder for artifacts storage
        run: mkdir /tmp/lighthouse-artifacts

      # It runs the lighthouse report for a provided url and create a HTML report in the specified directory
      # Action documentation: https://github.com/marketplace/actions/lighthouse-check#usage-standard-example
      - name: Run Lighthouse
        uses: foo-software/lighthouse-check-action@v2.0.5
        id: lighthouseCheck
        with: # See https://github.com/marketplace/actions/lighthouse-check#inputs for all options
          outputDirectory: /tmp/lighthouse-artifacts # Used to upload artifacts.
          emulatedFormFactor: all # Run LightHouse against "mobile", "desktop", or "all" devices
          urls: ${{ env.ZEIT_DEPLOYMENT_URL }}, ${{ env.ZEIT_DEPLOYMENT_URL }}/en, ${{ env.ZEIT_DEPLOYMENT_URL }}/fr, ${{ env.ZEIT_DEPLOYMENT_URL }}/en-US, ${{ env.ZEIT_DEPLOYMENT_URL }}/fr-FR

      # Upload HTML report create by lighthouse, could be useful
      - name: Upload artifacts
        uses: actions/upload-artifact@v1
        with:
          name: Lighthouse reports
          path: /tmp/lighthouse-artifacts

      # Using a pre-build action to make the action fail if your score is too low. It can be really interesting to track a low score on a commit
      # You can remove this action IF you don't want lighthouse to be a blocking point in your CI
      # This default values so you need to change them by your own criteria
      # Official documentation: https://github.com/foo-software/lighthouse-check-status-action
      - name: Handle Lighthouse Check results
        uses: foo-software/lighthouse-check-status-action@v1.0.1
        with:
          lighthouseCheckResults: ${{ steps.lighthouseCheck.outputs.lighthouseCheckResults }}
          minAccessibilityScore: "50"
          minBestPracticesScore: "50"
          minPerformanceScore: "30"
          minProgressiveWebAppScore: "50"
          minSeoScore: "50"

```

**Comment by Vadorequest** [comment]
Didn't see anything weird indeed. What's the content of your now.app.staging.json and now.app.production.json?

**Comment by Vadorequest** [comment]
Also, we released a few improvements today about GHA, you might want to check them out! https://github.com/UnlyEd/next-right-now/discussions/166#discussioncomment-103624

**Comment by samuelcastro** [comment]
The content is exactly like this https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-gcms-lcz-sty/now.customer1.staging.json and https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-gcms-lcz-sty/now.customer1.production.json plus my own scope and variables.

**Comment by samuelcastro** [comment]
> Also, we released a few improvements today about GHA, you might want to check them out! [#166 (comment)](https://github.com/UnlyEd/next-right-now/discussions/166#discussioncomment-103624)

Yeah I saw it, I'll merge it in asap.

**Comment by Vadorequest** [comment]
As long as the `name` is the same between both it should be fine. I assume you set `NEXT_PUBLIC_PROJECT_REF="app"` in both as well.

I don't see anything wrong. `yarn deploy:$(cat now.json | jq -r '.build.env.NEXT_PUBLIC_PROJECT_REF'):production:simple --token $ZEIT_TOKEN` from the production workflow should read the now.json (whether it's linked to prod or staging) and resolve the `NEXT_PUBLIC_PROJECT_REF` which should be "app" in both cases, and use the `deploy:app:production:simple` script that uses the `now.app.production.json` file to deploy. I don't get how it could deploy your production environment even if you linked now.json with your staging file.

I mean, no matter which file is linked by now.json, the production workflow should deploy using the now.app.production.json file, and the staging workflow should deploy with the now.app.staging.json file.

**Comment by samuelcastro** [comment]
> As long as the name is the same between both it should be fine. I assume you set NEXT_PUBLIC_PROJECT_REF="app" in both as well.

Yes they're correct.

I agree, this is weird, I'll keep investigating.

**Comment by samuelcastro** [comment]
Maybe this is related: https://github.com/vercel/vercel/issues/3099

**Comment by Vadorequest** [comment]
Good finding, definitely related to your issue.

I'm not sure how people "lock-in" a specific version of "now" when used through github actions, I noticed it always uses the latest without caring for what's specified in our own package.json.

Your workflow logs should print out what version is used, you can check there. 
I checked against a recent workflow and it uses `vercel@20.1.2` 🤔

**Comment by samuelcastro** [comment]
Same here: `Vercel CLI 20.1.2`

**Comment by Vadorequest** [comment]
Maybe try using the latest canary version `20.1.2-canary.5` and see if it solves the issue.

But... If Vercel does what I think it does then it'll still use 20.1.2. I think there is a way to lock down what version to use but don't know how.

**Comment by samuelcastro** [comment]
I'll try that tomorrow morning.

**Comment by samuelcastro** [comment]
Well, it seems to be working.

**Comment by samuelcastro** [comment]
The weird thing is that I'm still seeing `Vercel CLI 20.1.2` in the logs.

**Comment by Vadorequest** [comment]
Okay, thanks for the feedback.

I'm closing this, the issue is still there but the patch should be released in a non-canary release at some point in the future.
